composer/datasets/c4.py

```diff
@@ -0,0 +1,333 @@
+        microbatches = []
+        for idx in range(n_microbatches):
+            mb = {k: v[idx] for k, v in chunked.items()}
+            microbatches.append(mb)
+        return microbatches
+    else:
+        raise ValueError(f'Expected batch to be of type Dict[str, Tensor], but got {type(batch)}')
+
+
+@dataclass
+class C4DatasetHparams(DatasetHparams):
+    """Builds a DataSpec for the C4 (Colossal Cleaned CommonCrawl) dataset.
+
+    Parameters:
+        split (str): What split of the dataset to use. Either `train` or `validation`
+        max_samples (int): Max number of post-processed token samples, used to set epoch size of the IterableDataset.
+        tokenizer_name (str): The name of the HuggingFace tokenizer to preprocess text with.
+        max_seq_len (int): The max sequence length of each token sample.
+        group_method (str): How to group text samples into token samples. Either `truncate` or `concat`.
+        mlm (bool): Whether or not to use masked language modeling. (Default: `False`)
+        mlm_probability (float): If `mlm=True`, the probability that tokens are masked. (Default: `0.15`)
+        shuffle (bool): Whether to shuffle the samples in the dataset. Currently, shards are assigned deterministically to devices, but shuffling affects the order of shards (per-device), and the order of samples via (per-device) shuffle buffers (Default: `False`)
+        shuffle_buffer_size (int): If `shuffle=True`, samples are read into a buffer of this size (per-device), and randomly sampled from there to produce shuffled samples (Default: `10000`)
+        seed (int): If `shuffle=True`, what seed to use for shuffling operations. (Default: `5`)
+        drop_last (bool): Whether to drop the last samples for the last batch. (Default: `True`)
+    Returns:
+        A :class:`~composer.core.DataSpec` object
+    """
+
+    split: str = hp.optional("What split of the dataset to use. Either `train` or `validation`", default=None)
+    max_samples: int = hp.optional(
+        "Max number of post-processed token samples, used to set epoch size of the IterableDataset.", default=None)
+    tokenizer_name: str = hp.optional("The name of the HuggingFace tokenizer to preprocess text with.", default=None)
+    max_seq_len: int = hp.optional("The max sequence length of each token sample.", default=None)
+    group_method: str = hp.optional("How to group text samples into token samples. Either `truncate` or `concat`.",
+                                    default=None)
+    mlm: bool = hp.optional("Whether or not to use masked language modeling.", default=False)
+    mlm_probability: float = hp.optional("If `mlm=True`, the probability that tokens are masked.", default=0.15)
+    shuffle: bool = hp.optional(
+        "Whether to shuffle the samples in the dataset. Currently, shards are assigned deterministically to devices, but shuffling affects the order of shards (per-device), and the order of samples via (per-device) shuffle buffers.",
```

Comment:
Why would shuffling affect the order of the shards, rather than just shuffle within a shard?

Reply:
Sadly this is what the HF IterableDatset.shuffle function does, and there's no flag to turn off shard shuffling: https://huggingface.co/docs/datasets/package_reference/main_classes.html#datasets.IterableDataset.shuffle. 

However I could potentially move the shuffle op to the very end of the dataset definition, after `.take(samples_per_device)`. This would hypothetically stop the shard shufffling and mean that `max_samples` will always refer to the same subset regardless of seed. I think this is preferable...

... but note that `max_samples` is still affected by the number of devices! And that is impossible to fix... because if using 1 device the first 10 samples are all from shard1, but if using 2 devices the first 10 samples are (5 from shard1, 5 from shard2).